### PR TITLE
AWS Terraform Quick Start

### PR DESCRIPTION
**Pull request by travelton**
- Reorganizes the examples/aws/terraform directory to account for additional Terraform example
- Adds a simple "all in one" Teleport Cluster via Terraform scripting
- Updates docs to link to new directory structure.

### CODE REVIEW

File: docs/4.0/aws_oss_guide.md [hunk 1]
```diff
@@ -79,7 +79,7 @@ types of Teleport data:
 ![Setting Streams](img/aws/setting-stream.png)
 Setting Stream to `NEW IMAGE`
 
-For maintainability and ease of use, we recommend following our [Terraform example](https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/dynamo.tf)
+For maintainability and ease of use, we recommend following our [Terraform example](https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/ha-autoscale-cluster/dynamo.tf)
```
Comment (benarent) [inline docs/4.0/aws_oss_guide.md:82]: +1 points for backporting.

File: examples/aws/terraform/README.md [hunk 2]
```diff
@@ -1,68 +1,34 @@
...
 
-```bash
-# Set variables for Terraform
+In order to spin up AWS resources using these Terraform examples, you need the following software:
 
-# Region to run in - we currently have AMIs in the following regions:
-# ap-south-1,ap-northeast-2,ap-southeast-1,ap-southeast-2,ap-northeast-1,ca-central-1,eu-central-1,eu-west-1,eu-west-2
-# sa-east-1,us-east-1,us-east-2,us-west-1,us-west-2
-export TF_VAR_region="us-west-2"
+- terraform v0.12+ ([install docs](https://learn.hashicorp.com/terraform/getting-started/install.html))
```
Comment (benarent) [inline examples/aws/terraform/README.md:19]:
```suggestion
- terraform v0.12+ [install docs](https://learn.hashicorp.com/terraform/getting-started/install.html)
```

File: examples/aws/terraform/README.md [hunk 3]
```diff
@@ -1,68 +1,34 @@
...
+- awscli v1.14+ ([install docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html))
```
Comment (benarent) [inline examples/aws/terraform/README.md:20]:
```suggestion
- awscli v1.14+ [install docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
```

File: examples/aws/terraform/README.md [hunk 4]
```diff
@@ -1,68 +1,34 @@
...
+  - [Get Started](ha-autoscale-cluster/README.md)
 
-# Full absolute path to the license file for Teleport Enterprise or Pro.
-# This license will be copied into SSM and then pulled down on the auth nodes to enable Enterprise/Pro functionality
-export TF_VAR_license_path="/path/to/license"
+## How to get help
 
-# Route 53 zone to use, should be the zone registered in AWS, e.g. example.com
-export TF_VAR_route53_zone="example.com"
+If you're having trouble, check out our Discourse community. <https://community.gravitational.com>
```
Comment (benarent) [inline examples/aws/terraform/README.md:32]:
```suggestion
If you're having trouble, check out our [Discourse community](https://community.gravitational.com). 
```

File: examples/aws/terraform/ha-autoscale-cluster/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+# Email of your support org, used for Letsencrypt cert registration process.
+export TF_VAR_email="support@example.com"
+
+# Setup grafana password for "admin" user. Grafana will be served on https://cluster.example.com:8443 after install
+export TF_VAR_grafana_pass="CHANGE_THIS_VALUE"
+
+# (optional) Set to true to use ACM (Amazon Certificate Manager) to provision certificates rather than Letsencrypt
+# If you wish to use a pre-existing ACM certificate rather than having Terraform generate one for you, you can import it:
+# terraform import aws_acm_certificate.cert <certificate_arn>
+# export TF_VAR_use_acm="false"
```
Comment (benarent) [inline examples/aws/terraform/ha-autoscale-cluster/README.md:64]: I would move this up, to line 40, as ACM is easier and more AWS default ( if already bootstrapped )

File: examples/aws/terraform/starter-cluster/README.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+A series of systemd [units](https://github.com/gravitational/teleport/tree/master/assets/marketplace/files/system) bootstrap the instance, via several bash [scripts](https://github.com/gravitational/teleport/tree/master/assets/marketplace/files/bin).
+
+While this may not be sufficient for all use cases, it's a great proof-of-concept that you can fork and customize to your liking. Check out our AWS AMI [generation code](https://github.com/gravitational/teleport/tree/master/assets/marketplace) if you're interested in adapting this to your requirements.
+
+This Terraform example will configure the following AWS resources:
+
+- Teleport all-in-one (auth, node, proxy) single cluster ec2 instance
+- DynamoDB tables (cluster state, cluster events, ssl lock)
+- S3 bucket (session recording storage)
+- Route53 A record
```
Comment (benarent) [inline examples/aws/terraform/starter-cluster/README.md:20]:
```suggestion
- Route53 `A` record
```

File: examples/aws/terraform/starter-cluster/README.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+- Security Groups and IAM roles
+
+## Instructions
+
+### Build Requirements
+
+- terraform v0.12+ ([install docs](https://learn.hashicorp.com/terraform/getting-started/install.html))
```
Comment (benarent) [inline examples/aws/terraform/starter-cluster/README.md:27]:
```suggestion
- terraform v0.12+ [install docs](https://learn.hashicorp.com/terraform/getting-started/install.html)
```

File: examples/aws/terraform/starter-cluster/README.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+- awscli v1.14+ ([install docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html))
```
Comment (benarent) [inline examples/aws/terraform/starter-cluster/README.md:28]:
```suggestion
- awscli v1.14+ [install docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
```

**Review by benarent — APPROVED** [review]
While a large PR, I'm really digging the smaller TF example. I've not tested it locally but it looks good.

**Comment by benarent** [comment]
@travelton Just cleaning up older PRs. I think this is in a good state? Feel free to review my suggestions, I can give you a demo of how to merge them this week.

**Comment by travelton** [comment]
> @travelton Just cleaning up older PRs. I think this is in a good state? Feel free to review my suggestions, I can give you a demo of how to merge them this week.

@benarent Tweaks completed and brought up to date with master. Ready for official review. I use this TF script often, so I'm confident it works. As for the large amount of documentation changes... I will want to deploy this to a staging environment to perform a final validation. Can you help with that?

File: examples/aws/terraform/README.md [hunk 9]
```diff
@@ -1,50 +1,54 @@
...
-Use these examples as possible deployment patterns suggested by Teleport developers.
+- A simple starter Teleport cluster to, quickly and cost-effectively, demo or POC Teleport on a single node (auth, proxy, and node processes running on one t3.nano ec2 instance).
+- A production worthy high-availability auto-scaling Teleport Cluster. This cluster makes use of several AWS technologies, provisioned and configured using Terraform.
 
+<<<<<<< HEAD
+If you are planning on using our Terraform example in production, please reference the high-availability auto-scaling Teleport Cluster for best practices. Our Production Guide outlines in-depth details on how to run Teleport in production.
+=======
 The scripts set up Letsencrypt certificates using DNS-01 challenge. This means that users have to control the DNS zone
 via Route 53. ACM can optionally be used too, but Route 53 integration is still required.
+>>>>>>> master
```
Comment (benarent) [inline examples/aws/terraform/README.md:13]: FYI looks like some git issues here.

Comment (travelton) [reply examples/aws/terraform/README.md]: Thanks! Fixed.

**Comment by benarent** [comment]
retest this please

**Comment by benarent** [comment]
retest this please

**Comment by benarent** [comment]
retest this please

**Comment by benarent** [comment]
It looks like Jenkins is a detached state?

Since Drone is showing a 🟢 ✔️ , are we ok using my Admin privileges to merge this? Or should be fix this jenkins issue @webvictim 

```log
Note: checking out '16732679581d29fd486a5e667fc88b36fe20422f'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at 1673267... Fix README conflicts.
Submodule 'e' (git@github.com:gravitational/teleport.e.git) registered for path 'e'
Cloning into 'e'...
fatal: reference is not a tree: ec20619df05c362702855184296ced13ba59354a
Unable to checkout 'ec20619df05c362702855184296ced13ba59354a' in submodule path 'e'
Build step 'Execute shell' marked build as failure
Setting status of 16732679581d29fd486a5e667fc88b36fe20422f to FAILURE with url https://jenkins.gravitational.io/job/Teleport/7025/ and message: 'Build finished. '
Using context: Teleport
Finished: FAILURE
```

**Comment by webvictim** [comment]
I think that the failure is because the `e` submodule ref in Teleport master was broken for a while.
@travelton You'll need to rebase this PR against the current state of `master`. Something like this on your branch should do it:

```console
git fetch origin master
git rebase origin master
```

Unfortunately there are likely to be a few merge conflicts because it's been a couple of months since the PR was opened. Once you've fixed them up you can push the branch again and Jenkins should be happy.

**Comment by travelton** [comment]
@webvictim 

I brought this branch up to date with master today. It's currently just 4 commits behind master. Hmm.

**Comment by webvictim** [comment]
@travelton How long ago? https://github.com/gravitational/teleport/commit/148d02a43019be9e8b5acaf4769da3cdcd53c642 is the commit that fixed the e ref earlier today. I think that it was _also_ broken when you originally created the PR. Very bad luck 😁

**Comment by benarent** [comment]
In this case I think I'm ok to use my admin privileges to merge since it isn't on a core testing path. I can do this in the AM

**Comment by travelton** [comment]
Oh sorry! I didn't realize the fix went in today. Ok yep, I'm missing that commit. Will rebase.

**Pull request merged**